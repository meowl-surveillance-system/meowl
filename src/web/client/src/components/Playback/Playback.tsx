import React, { Component } from "react";
import { Container, Typography } from "@material-ui/core";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListSubheader from "@material-ui/core/ListSubheader";

import ResponsivePlayer from "../ResponsivePlayer/ResponsivePlayer";

interface Props {}
interface State {
  tmpUrl: string;
  url: string;
  cameraIds: Array<string>;
  cameraIdsDict: Record<string, Array<string>>;
}

export default class Playback extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      tmpUrl: "",
      url: "",
      cameraIds: [],
      cameraIdsDict: {},
    };
  }

  // Map cameraId to streamIds
  componentDidMount() {
    fetch(`/api/getCameraIds`)
      .then((res) => res.json())
      .then((cameraIds) => this.setState({ cameraIds: cameraIds }))
      .then(() =>
        this.state.cameraIds.map((cameraId, index) => {
          fetch(`/api/getStreamIds/${cameraId}`)
            .then((res) => res.json())
            .then((streamIds) =>
              this.setState({
                cameraIdsDict: {
                  ...this.state.cameraIdsDict,
                  [cameraId]: streamIds,
                },
              }),
            );
        }),
      );
  }

  // Render the nested list of cameraIds and streamIds
  renderList = () => {
    return Object.keys(this.state.cameraIdsDict).map((cameraId, index) => {
      return (
        <List
          key={index}
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">{cameraId}</Typography>
            </ListSubheader>
          }
        >
          {this.renderStreamIds(this.state.cameraIdsDict[cameraId])}
        </List>
      );
    });
  };

  renderStreamIds = (streamIds: Array<string>) => {
    return streamIds.map((streamId, index) => {
      return (
        <ListItem
          button
          key={index}
          onClick={() => this.retrieveVideo(streamId)}
        >
          {streamId}
        </ListItem>
      );
    });
  };

  urlSubmit = (): void => {
    this.setState({ url: this.state.tmpUrl });
    this.setState({ tmpUrl: "" });
  };

  retrieveVideo = (streamId: string): void => {
    fetch(`/api/getVideo/${streamId}`)
      .then((res) => res.blob())
      .then((blob) => {
        const vidUrl = URL.createObjectURL(new Blob([blob]));
        this.setState({ url: vidUrl });
      });
  };

  render() {
    return (
      <Container>
        <List
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">My Streams</Typography>
            </ListSubheader>
          }
        >
          {this.renderList()}
        </List>
        <ResponsivePlayer url={this.state.url} controls={true} />
      </Container>
    );
  }
}
