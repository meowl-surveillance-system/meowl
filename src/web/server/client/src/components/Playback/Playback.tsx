import React, { ComponentState } from 'react';
import ReactPlayer from 'react-player';

type myState = {
  tmpUrl: string;
  url: string;
  startTime: string;
  vidId: string;
}

export default class Playback extends React.Component {
  state: myState = {
    tmpUrl: "",
    url: "",
    startTime: "",
    vidId: ""
  }

  handleChange = (field: string) => (event: React.FormEvent<HTMLInputElement>): void => {
    event.preventDefault();
    this.setState({ [field]: event.currentTarget.value} as ComponentState);
  };

  handleSubmit = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    this.setState({ url: this.state.tmpUrl });
    this.setState({ tmpUrl: "" });
  };

  retrieveVideo = (event: React.FormEvent<HTMLFormElement>): void => {
    event.preventDefault();
    fetch(`http://localhost:8081/api/getVideo/?start=${this.state.startTime}&id=${this.state.vidId}`)
    .then(res => res.blob())
    .then(blob => {
      const vidUrl = URL.createObjectURL(new Blob([blob]));
      this.setState({ url: vidUrl });
    })
      
  }

  render() {
    return(
      <div>
        <form onSubmit={this.handleSubmit} data-testid="form">
          <label>Video URL</label>
          <input type="text" value={this.state.tmpUrl} onChange={this.handleChange('tmpUrl')} data-testid="url" />
          <button type="submit">Enter URL</button>
        </form>

        <form onSubmit={this.retrieveVideo} data-testid="retrieve">
          <p>Video Retrieval:</p>
          <label>Start Time</label>
          <input type="text" value={this.state.startTime} onChange={this.handleChange('startTime')} data-testid="" />
          <label>Video Id</label>
          <input type="text" value={this.state.vidId} onChange={this.handleChange('vidId')} data-testid="" />
          <button type="submit">Retrieve Video</button>
        </form>
        <ReactPlayer url={this.state.url} controls={true} />
      </div>
    );
  }
}