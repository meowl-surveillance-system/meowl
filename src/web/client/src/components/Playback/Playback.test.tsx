import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import Playback from './Playback';

describe('Playback component', () => {
    test('renders Playback component', () => {
        const { container } = render(<Playback />);
        expect(container.firstChild).toMatchSnapshot();
    });

    test('change input and form submission', () => {
        render(<Playback />);
        const urlInput = screen.getByTestId('url') as HTMLInputElement;
        const url = 'testURL';
        fireEvent.change(urlInput, {target: {value: url}})
        expect(urlInput.value).toEqual(url);
        const form = screen.getByTestId('form') as HTMLFormElement;
        fireEvent.submit(form);
        expect(urlInput.value).toEqual('');
    });
});