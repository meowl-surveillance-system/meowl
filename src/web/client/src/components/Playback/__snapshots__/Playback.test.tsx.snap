// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Playback component renders Playback component 1`] = `
<div>
  <form
    data-testid="form"
  >
    <input
      data-testid="url"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Enter URL
    </button>
  </form>
  <div
    style="width: 640px; height: 360px;"
  />
</div>
`;
